app/validators/answer_code_validator.py

```diff
@@ -0,0 +1,110 @@
+        self.validate_missing_answer_id()
+        self.validate_missing_answer_codes()
+        self.validate_answer_codes_at_option_level()
+        return self.errors
+
+    def validate_data_version(self):
+        if self.data_version != "0.0.3":
+            self.add_error(self.INCORRECT_DATA_VERSION_FOR_ANSWER_CODES)
+
+    def validate_duplicate_answer_codes(self):
+        duplicates = find_duplicates(self.codes)
+
+        if len(duplicates) > 0:
+            self.add_error(self.DUPLICATE_ANSWER_CODE_FOUND, duplicates=duplicates)
+
+    def validate_missing_answer_id(self):
+        all_answer_ids = list(self.questionnaire_schema.answers_with_context)
+
+        for answer_code_id in self.answer_codes_answer_ids:
+            if answer_code_id not in all_answer_ids:
+                self.add_error(
+                    self.ANSWER_CODE_ANSWER_ID_NOT_FOUND_IN_SCHEMA,
+                    answer_id=answer_code_id,
+                )
+
+    def validate_missing_answer_codes(self):
+        all_answer_ids = list(self.questionnaire_schema.answers_with_context)
+
+        for answer_id in all_answer_ids:
+            if answer_id not in self.answer_codes_answer_ids:
+                self.add_error(self.MISSING_ANSWER_CODE, answer_id=answer_id)
+
+    def validate_answer_codes_at_option_level(self):
+        answers_with_context = self.questionnaire_schema.answers_with_context
+
+        for answer_id in answers_with_context:
+            answer = answers_with_context[answer_id]
+
+            if "options" in answer["answer"]:
+                values = []
+                values.extend(option["value"] for option in answer["answer"]["options"])
+
+                answer_codes_for_options = [
+                    answer_code
+                    for answer_code in self.answer_codes
+                    if answer_code["answer_id"] == answer_id
+                ]
+
+                if len(values) != len(answer_codes_for_options) and all(
+                    "value" in answer_code for answer_code in answer_codes_for_options
+                ):
+                    self.add_error(
+                        self.ANSWER_CODE_MISSING_FOR_ANSWER_OPTIONS,
+                        answer_options=answer["answer"]["options"],
+                        answer_codes_for_options=answer_codes_for_options,
+                    )
+
+                for answer_code in answer_codes_for_options:
+                    if answer_code.get("value") and answer_code["value"] not in values:
+                        self.add_error(
+                            self.INCORRECT_VALUE_FOR_ANSWER_CODE_WITH_ANSWER_OPTIONS,
+                            value=answer_code["value"],
+                            answer_codes_for_options=answer_codes_for_options,
+                        )
+
+                if len(answer_codes_for_options) != 1 and any(
+                    "value" not in answer_code
+                    for answer_code in answer_codes_for_options
+                ):
```

Comment:
This doesn't cover the case where you have 1 code for options but it also defines "answer_value".

Reply:
Updated the logic here and added another test 👍